src/NServiceBus.Core/Sagas/Sagas.cs

```diff
@@ -104,6 +104,13 @@ static bool IsCompatible(Type t, Type source)
 
     static bool IsTypeATimeoutHandledByAnySaga(Type type, IEnumerable<Type> sagas)
     {
+        // MakeGenericType() throws an exception if passed a ref struct type
+        // Messages cannot be ref struct types
```

Comment:
this is not the correct fix. types that do not match the convention should never get to this point. the bug is that RouteToEndpoint has a side effect of causing non message types to be treated as message types

Reply:
> the bug is that RouteToEndpoint has a side effect of causing non message types to be treated as message types

In order to determine if any given type is a message type or not, we have to ask the conventions subsystem. [Here is where that is happening](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Routing/AssemblyRouteSource.cs#L23).

If you have a saga in your endpoint, [the Sagas feature adds a message convention](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Sagas/Sagas.cs#L25) to say that [anything handled by a saga IS a message type](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Sagas/Sagas.cs#L105-L111), even if it matches no other conventions. This is how you can have timeout messages that do not conform to any registered convention.

To make that determination we use the type in question to create the two handler type interfaces that sagas might implement (`IHandleMessages<>` and `IHandleTimeouts<>`) and then check to see if any sagas implement those. 

So unfortunately we cannot say that this type is not a message type without executing this code. We could (and possibly will) move the check even earlier in the conventions pipeline but this change is the least invasive one we could devise for a patch release.

Reply:
@mikeminutillo ahh. fair enough. thanks for the detailed explanation